File: frontend/src/api/services/host.ts
import { get, post, del } from '../client';
import { HostApi, LabelApi } from '../constants';

import type { LabelInfo } from './label';

export interface DiskInfo {
  id: number;
  diskSpaceAvailable: number;
  totalDiskSpace: number;
  percentDiskSpaceAvailable: string;
  encrypted: boolean;
  createdAt: string;
}

export interface MyHostInfo {
  labelName: string;
  labelId: number;
  hosts: {
    hostId: number;
    hostName: string;
    hostIp: string;
  }[];
}

export interface HostInfo {
  id: number;
  name: string;
  hostServerUrl: string;
  uuid: string;
  primaryMac: string;
  label: LabelInfo[];
  operatingSystem: string;
  version: string;
  arch: string;
  platform: string;
  kernelVersion: string;
  physicalMemory: string;
  cpuBrand: string;
  cpuType: string;
  cpuPhysicalCores: number;
  cpuLogicalCores: number;
  cpuSockets: number;
  cpuMicrocode: string;
  hardwareModel: string;
  hardwareVersion: string;
  hardwareVendor: string;
  boardModel: string;
  boardVersion: string;
  boardVendor: string;
  boardSerial: string;
  totalUptimeSeconds: string;
  startTime: string;
  fetchTime: string;
  createdAt: string;
  updatedAt: string;
  disk: DiskInfo[];
  hostUser: any[];
  software: any[];
  metrics?: MetricsInfo;
}

export interface CpuMetrics {
  usagePercent: number;
  load1m: number;
  load5m: number;
  load15m: number;
}

export interface MemoryMetrics {
  totalBytes: number;
  usedBytes: number;
  freeBytes: number;
  availableBytes: number;
  usagePercent: number;
}

export interface DiskVolumeMetrics {
  mountPoint: string;
  deviceName: string;
  totalBytes: number;
  usedBytes: number;
  freeBytes: number;
  usagePercent: number;
  fsType: string;
}

export interface DiskMetrics {
  availableBytes: number;
  totalBytes: number;
  usagePercent: string;
  readRate: number;
  writeRate: number;
  volumes: DiskVolumeMetrics[];
}

export interface NetworkInterfaceMetrics {
  name: string;
  macAddress: string;
  ipv4Address: string;
  totalRecvBytes: number;
  totalSentBytes: number;
  recvRate: number;
  sendRate: number;
}

export interface NetworkMetrics {
  recvRate: number;
  sendRate: number;
  interfaces: NetworkInterfaceMetrics[];
}

export interface MetricsInfo {
  collectedAt: number;
  hostId: number;
  cpu: CpuMetrics;
  memory: MemoryMetrics;
  disk: DiskMetrics;
  network: NetworkMetrics;
}

export interface HostListResponse {
  count: number;
  list: HostInfo[];
}

export interface MyHostListResponse {
  count: number;
  list: MyHostInfo[];
}

/**
 * 主机管理
 */
const hostService = {
  /** 获取主机列表 */
  getHosts() {
    return get<HostListResponse>(HostApi.List);
  },

  /** 获取带监控主机列表 */
  getHostsWithMetrics() {
    return get<HostListResponse>(`${HostApi.List}?withMetrics=true`);
  },

  /** 获取我的主机列表 */
  getMyHosts() {
    return get<MyHostListResponse>(HostApi.GetMine);
  },

  /** 获取主机详情 */
  getHostDetail(hostId: number) {
    return get<HostInfo>(HostApi.ById.replace(':id', hostId.toString()));
  },

  /** 更新主机名称 */
  updateHostName(params: { hostname: string; hostServerUrl: string }) {
    return post<HostInfo>(HostApi.Rename, params);
  },

  /** 删除主机 */
  deleteHosts(hostIds: number[]) {
    return del<void>(HostApi.Delete, { hostIds });
  },

  /** 分配标签 */
  assignLabels(params: { hostId: number; labelIds: number[] }) {
    return post<HostInfo>(LabelApi.AssignToHost, params);
  },

  /** 获取新增主机命令 */
  getCreateHostCommand(osFamily: string) {
    return get<{ command: string }>(`${HostApi.Install}?osFamily=${osFamily}`);
  },

  /** 搜索主机 */
  searchHosts(keyword: string) {
    return get<HostListResponse>(`${HostApi.Search}?keyword=${keyword}`);
  },
};

export default hostService;

File: api/app/schemas/host.py
from __future__ import annotations

from pydantic import BaseModel, Field
from datetime import datetime

def to_camel(string: str) -> str:
    """将下划线命名转换为驼峰命名"""
    components = string.split('_')
    return components[0] + ''.join(x.capitalize() for x in components[1:])

# 基础模型类，包含通用配置
class BaseSchema(BaseModel):
    class Config:
        from_attributes = True
        alias_generator = to_camel
        populate_by_name = True

# 注解模型（层级格式）
class AnnotationSchemaForHost(BaseSchema):
    id: int
    name: str  # 如 "server/env"
    value: str  # 如 "prod"
    namespace: str | None = None  # 如 "server"
    key: str  # 如 "env"
    created_at: datetime
    updated_at: datetime

class HostUserSchema(BaseSchema):
    id: int
    host_id: int
    username: str
    group_name: str | None = None
    shell: str | None = None

class DiskSchema(BaseSchema):
    id: int
    host_id: int
    disk_space_available: float
    total_disk_space: float
    percent_disk_space_available: str
    encrypted: bool = False

class SoftwareSchema(BaseSchema):
    id: int
    host_id: int
    name: str
    version: str
    type: str

class HostSchema(BaseSchema):
    id: int
    created_at: datetime
    updated_at: datetime
    
    # 基本信息 - 与Go版本对应
    name: str
    operating_system: str | None = None
    host_server_url: str | None = None
    
    # 硬件信息 - 与Go版本对应  
    cpu_brand: str | None = None
    physical_memory: str | None = None
    kernel_version: str | None = None
    
    # 时间信息 - 与Go版本对应
    fetch_time: datetime | None = None
    start_time: datetime | None = None
    
    # 其他字段保持兼容
    status: int = 0
    agent: str | None = None
    osquery_host_id: str | None = None
    osquery_version: str | None = None
    platform_like: str | None = None
    platform: str | None = None
    arch: str | None = None
    instance_id: str | None = None
    cpu_type: str | None = None
    cpu_logical_cores: str | None = None
    cpu_physical_cores: str | None = None
    cpu_sockets: str | None = None
    cpu_subtype: str | None = None
    cpu_microcode: str | None = None
    board_model: str | None = None
    board_serial: str | None = None
    board_vendor: str | None = None
    board_version: str | None = None
    hardware_model: str | None = None
    hardware_serial: str | None = None
    hardware_vendor: str | None = None
    hardware_version: str | None = None
    computer_name: str | None = None
    uuid: str | None = None
    total_uptime_seconds: str | None = None
    build: str | None = None
    major: str | None = None
    minor: str | None = None
    patch: str | None = None
    version: str | None = None
    org: str | None = None
    primary_ip: str | None = None
    primary_mac: str | None = None
    public_ip: str | None = None
    country: str | None = None
    city: str | None = None

class HostSchemaWithRelations(HostSchema):
    annotations: list[AnnotationSchemaForHost] = []
    disk: list[DiskSchema] = []
    label: list[AnnotationSchemaForHost] = []  # 兼容Go版本的label字段名
    metrics: dict | None = None  # 监控指标数据，可选

# 请求和响应模型
class HostListRequest(BaseModel):
    page: int = 1
    limit: int = 20
    key: str | None = None
    annotation_ids: list[int] | None = None
    logic: str = "and"  # "and" or "or"
    with_metrics: bool = False


class HostListResponse(BaseModel):
    list: list[HostSchemaWithRelations]  # 使用带关联数据的Schema
    count: int